test/e2e_node/image_list.go

```diff
@@ -151,23 +155,62 @@ func PrePullAllImages() error {
 	}
 	images := framework.ImageWhiteList.List()
 	klog.V(4).Infof("Pre-pulling images with %s %+v", puller.Name(), images)
-	for _, image := range images {
-		var (
-			err    error
-			output []byte
-		)
-		for i := 0; i < maxImagePullRetries; i++ {
-			if i > 0 {
-				time.Sleep(imagePullRetryDelay)
-			}
-			if output, err = puller.Pull(image); err == nil {
-				break
+
+	imageCh := make(chan int, len(images))
+	for i := range imageCh {
+		imageCh <- i
+	}
+	close(imageCh)
+
+	pullErrs := make([]error, len(images))
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	parallelImagePullCount := maxParallelImagePullCount
+	if len(images) < parallelImagePullCount {
+		parallelImagePullCount = len(images)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(parallelImagePullCount)
+	for i := 0; i < parallelImagePullCount; i++ {
+		go func() {
+			defer wg.Done()
+
+			for i := range imageCh {
+				var (
+					pullErr error
+					output  []byte
+				)
+				for retryCount := 0; retryCount < maxImagePullRetries; retryCount++ {
+					select {
+					case <-ctx.Done():
+						return
+					default:
+					}
+
+					if retryCount > 0 {
+						time.Sleep(imagePullRetryDelay)
+					}
+					if output, pullErr = puller.Pull(images[i]); pullErr == nil {
+						break
+					}
+					klog.Warningf("Failed to pull %s as user %q, retrying in %s (%d of %d): %v",
+						images[i], usr.Username, imagePullRetryDelay.String(), retryCount+1, maxImagePullRetries, pullErr)
+				}
+				if pullErr != nil {
+					klog.Warningf("Could not pre-pull image %s %v output: %s", images[i], pullErr, output)
+					pullErrs[i] = pullErr
+					cancel()
+					return
+				}
 			}
-			klog.Warningf("Failed to pull %s as user %q, retrying in %s (%d of %d): %v",
-				image, usr.Username, imagePullRetryDelay.String(), i+1, maxImagePullRetries, err)
-		}
+		}()
+	}
+
+	wg.Wait()
+	for _, err := range pullErrs {
```

Comment:
Final request - can we aggregate all of the `non-nil` `pullErrors` into a single error, instead of just returning the first error? Thanks

Reply:
Aggregate all pullErrors into a single error, and If there are nothing errors in the pullErrors slice, the  utilerrors.NewAggregate function will return nil. PTAL